### PR TITLE
Windows: Graph driver implementation

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@swernli

This PR is part of the proposal described in docker/docker issue 10662 to port the docker daemon to Windows. This is the final piece of the jigsaw which is missing from master not yet PR'd - the Windows graph driver.

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks, @jhowardmsft !

File: daemon/daemon_windows.go [hunk 1]
```diff
@@ -6,11 +6,15 @@ import (
 	"runtime"
 	"syscall"
 
+	"github.com/Sirupsen/logrus"
 	"github.com/docker/docker/daemon/graphdriver"
+	_ "github.com/docker/docker/daemon/graphdriver/vfs"
```
Comment (crosbymichael) [inline daemon/daemon_windows.go]: why is this imported?

**Comment by crosbymichael** [comment]
LGTM

please rebase

File: daemon/daemon_windows.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/daemon_windows.go]: You're right - not needed as already imported through daemon.go. Will remove.

**Comment by jhowardmsft** [comment]
@crosbymichael Thanks for reviewing. Rebased

**Comment by jfrazelle** [comment]
@jhowardmsft when can we start running the daemon locally and unskip a bunch of tests

File: daemon/graphdriver/windows/windows.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+// Diff produces an archive of the changes between the specified
+// layer and its parent layer which may be "".
+func (d *WindowsGraphDriver) Diff(id, parent string) (arch archive.Archive, err error) {
+	return nil, fmt.Errorf("The Windows graphdriver does not support Diff()")
+}
+
+// Changes produces a list of changes between the specified layer
+// and its parent layer. If parent is "", then all changes will be ADD changes.
+func (d *WindowsGraphDriver) Changes(id, parent string) ([]archive.Change, error) {
```
Comment (jfrazelle) [inline daemon/graphdriver/windows/windows.go]: will these be something that are implemented later

**Comment by jfrazelle** [comment]
LGTM

File: daemon/graphdriver/windows/windows.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: Maybe down the road, it's still a work in progress.

File: graph/graph.go [hunk 3]
```diff
@@ -37,6 +37,22 @@ type Graph struct {
...
+	Created         time.Time         `json:"created"`
+	Container       string            `json:"container,omitempty"`
+	ContainerConfig runconfig.Config  `json:"container_config,omitempty"`
+	DockerVersion   string            `json:"docker_version,omitempty"`
+	Author          string            `json:"author,omitempty"`
+	Config          *runconfig.Config `json:"config,omitempty"`
+	Architecture    string            `json:"architecture,omitempty"`
+	OS              string            `json:"os,omitempty"`
+	Size            int64
+	graph           Graph
```
Comment (stevvooe) [inline graph/graph.go]: We just got done decoupling this object from the graph driver... this is pretty disappointing. This is an API object coupled to internal storage formats.

Comment (stevvooe) [inline graph/graph.go]: Also, `Graph` is a concrete type. Embedding this makes no sense.

File: daemon/graphdriver/windows/windows.go [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+	return hcsshim.CopyLayer(d.info, sourceId, id, parentLayerPaths)
+}
+
+func (d *WindowsGraphDriver) LayerIdsToPaths(ids []string) []string {
+	var paths []string
+	for _, id := range ids {
+		path, err := d.Get(id, "")
+		if err != nil {
+			logrus.Debug("LayerIdsToPaths: Error getting mount path for id", id, ":", err.Error())
```
Comment (stevvooe) [inline daemon/graphdriver/windows/windows.go]: Why not return the error here?

File: daemon/graphdriver/windows/windows.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	logrus.Debugf("WindowsGraphDriver ApplyDiff: Untar time: %vs", time.Now().UTC().Sub(start).Seconds())
+
+	return
+}
+
+// DiffSize calculates the changes between the specified layer
+// and its parent and returns the size in bytes of the changes
+// relative to its base filesystem directory.
+func (d *WindowsGraphDriver) DiffSize(id, parent string) (size int64, err error) {
```
Comment (stevvooe) [inline daemon/graphdriver/windows/windows.go]: Completely unused.

File: graph/graph_windows.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+	return
+}
+
+// storeImage stores file system layer data for the given image to the
+// graph's storage driver. Image metadata is stored in a file
+// at the specified root directory.
+func (graph *Graph) storeImage(img *Image, layerData archive.ArchiveReader, root string) (err error) {
+
+	if wd, ok := graph.driver.(*windows.WindowsGraphDriver); ok {
```
Comment (stevvooe) [inline graph/graph_windows.go]: These two code blocks are nearly identical.

**Comment by stevvooe** [comment]
I'm a little worried about the code quality in this merge. I have a few issues summarized here:

- WindowsGraphDriver adds several methods that just repeat others:
	- `LayerIdsToPaths` -> `ProtoDriver.Get`
	- `Info` is unused externally, no need to define it
	- `Import`/`Export` -> `ApplyDiff`/`Diff`
- moves image.Image to graph package for no good reason
- many poor coding practices for Go (unnecessary interface definition, elided error returns)
- very light method documentation
- how are windows images actually going to be stored? Is tar sufficient? Are they building a new tar protocol that we'll have to support in the future?

My main concern here is that several interfaces a basically circumvented or ignored, making future maintenance impossible.

File: graph/graph_windows.go (hunk 6 above)
Comment (jhowardmsft) [inline graph/graph_windows.go]: Yes, but for the sake of code clarity to make is absolutely clear what is 'production' code and what is for development only using vfs, this is a lot cleaner.

Comment (stevvooe) [inline graph/graph_windows.go]: @jhowardmsft I'm sorry but this is not clear at all. The comment below refers to the `VFS` driver, which is just a single driver among many. We've also placed the "production" branch in an else block, which makes it seem like it is not the main flow.

The docker code base is not really the place to have "production" and "development" branches, even for experimental features. Even if it is experimental, the code quality *must* remain high.

Comment (jhowardmsft) [inline graph/graph_windows.go]: Stephen - I think you're missing the point of this code block. None of this is "experimental". It is intentionally possible to start the Windows daemon using vfs to allow the daemon to start without relying on Windows containers. This has been discussed several times on IRC. vfs cannot support actually running Windows containers as they rely on the Win32 windows filter driver to operate. However vfs is extremely useful as it is a "work anywhere" type of driver which will sit on top of NTFS. The else block here is the vfs path, not the Windows filter path, and the code here is near enough identical (actually is identical) to the unix code path. So I can't see that you can state code quality in this particular instance isn't high, unless you are also saying the unix code quality isn't high???

Clearly it is not possible to use any other graph drivers apart from the Windows driver or VFS in the Windows daemon - no Linux specific graphdriver (aufs, btrfs, devmapper, overlay, zfs and so on) could possibly work on top of NTFS in Windows. VFS is unique in being file system agnostic - it is a copy-style diff driver only.

**Comment by jhowardmsft** [comment]
Answering comments separately - some will come from @swernli. 

"Info is unused externally, no need to define it". Not so, it is used in three places in container_windows.go and once in daemon_windows.go so it is necessary to define it and be available externally.

**Comment by jhowardmsft** [comment]
"•how are windows images actually going to be stored? Is tar sufficient? Are they building a new tar protocol that we'll have to support in the future?". Stored for what? For transport? When in use? The whole point of platform specific graph drivers is that the format can be opaque to docker, so it shouldn't matter.

**Comment by jhowardmsft** [comment]
"•WindowsGraphDriver adds several methods that just repeat others: ◦  LayerIdsToPaths  ->  ProtoDriver.Get " LayerIdsToPath takes and array and returns an array while Get only acts on one ID at a time (and requires a put). LayerIdsToPath is essentially a helper method to avoid having to have each of the callers re-implement the pattern of iterating through the IDs, call Get and Put on each, and compiling the resulting list.

**Comment by jhowardmsft** [comment]
"•very light method documentation". I have to be honest, this is pretty consistent with the rest of the docker code base. But am happy to add more comments if it would make the code clearer.

**Comment by stevvooe** [comment]
The main issue here is that the `GraphDriver` interface has been circumvented and made it so the `Graph` must deal with both the `GraphDriver` interface and the `WindowsGraphDriver` interface. Everywhere there is a type switch, the behavior should be hidden by a driver method call. Otherwise, the use of a driver interface is pointless. If the driver interface is not sufficient, propose a change.

We have been trying to clean up the interface to `Graph` for some time and this interface bloat does not serve to enhance the situation. 

Furthermore, while discussions on IRC are important, we should at least have documentation about what the actual plan for windows support looks like. The data format is _not_ at all opaque to docker. Docker understands layers and exports a transport format that directly exposes these layers. There is nothing opaque about this situation. Understanding how this will work for windows is very important.

We are still in the process of making v2 image distribution work and additional functionality on the graph and driver interface may serve to slow these efforts. If we can see what the full plan is, we may be able to provide advice or direction that may make things much more straightforward.

File: daemon/container_windows.go [hunk 7]
```diff
@@ -102,6 +106,26 @@ func populateCommand(c *Container, env []string) error {
...
+		var ids []string
+
+		if img, err = c.daemon.graph.Get(c.ImageID); err != nil {
+			return fmt.Errorf("Failed to graph.Get on ImageID %s - %s", c.ImageID, err)
+		}
+		if ids, err = c.daemon.graph.ParentLayerIds(img); err != nil {
+			return fmt.Errorf("Failed to get parentlayer ids %s", img.ID)
+		}
+		layerPaths = wd.LayerIdsToPaths(ids)
+		layerFolder = filepath.Join(wd.Info().HomeDir, filepath.Base(c.ID))
```
Comment (stevvooe) [inline daemon/container_windows.go]: Clarifying my point on `Info`: the type assertion has been made. No need to have a special interface for this declared. Just access the struct field.

**Comment by icecrime** [comment]
I'm receiving concerns about this PR from several people. I haven't looked into it myself, but just to understand: @stevvooe are we talking about things that can be fixed in a subsequent PR, or would you advocate a revert?

Of all the Windows specific PR, this is not the biggest one, I'm sure we can figure something out.

File: daemon/container_windows.go (hunk 7 above)
Comment (swernli) [inline daemon/container_windows.go]: This must be my misunderstanding then.  When I wrote this, I tried to add the struct to the interface, and found that only methods can be added to an interface.  When I tried to access the struct directly without a method, it didn't work, so I was under the impression it was a limitation of Go that required me to have a method to access the struct.  If that's not the case, then it's easy enough to adjust the pattern to whatever allows us to get the information we need and remove the method.

Just for my edification, is there any reason why having a method for accessing the struct is bad?

**Comment by stevvooe** [comment]
@icecrime These can all be fixed in a separate PR. We just need to make sure that these abstractions are held or modified.

@jhowardmsft I'm willing to supply some time to help get this into shape.

**Comment by dmp42** [comment]
@jhowardmsft @swernli can you reach out so we figure this together?
cc @vbatts

**Comment by vbatts** [comment]
+1

**Comment by jhowardmsft** [comment]
Yup - swernli is OOF today - sent an invite for 11AM PST on Monday. @stevvooe I don't have your email address, hopefully @icecrime or @dmp42 will forward the invite.

**Comment by stevvooe** [comment]
@jhowardmsft I got your email. Thank you!

**Pull request merged**